[이상운] Sprint 5

unzzang opened the pull request:
# 미션 목표

- [x] 타입스크립트 마이그레이션
- [x] 타입스크립트 개발 환경 세팅
- [x] Layered Architecture 적용

# 기본 요구 사항

- [x] 스프린트 미션 4 구현이 완료된 상태에서 진행
- [x] 타입스크립트 마이그레이션 진행, 
	- [x] 이전 미션에서 구현하지 못한 부분이 있다면 추가로 구현

## 프로젝트 세팅

- [x] `tsconfig.json` 파일 생성 및 옵션 설정
- [x] 필요한 npm script 설정 (예: 빌드 및 개발 서버 실행 명령어)

## 타입스크립트 마이그레이션

- [x] 기존 Express.js 프로젝트를 타입스크립트 프로젝트로 마이그레이션
- [x] 필요한 타입 패키지 설치
- [x] `any` 타입의 사용은 최소화
- [x] 복잡한 객체 구조나 배열 구조를 가진 변수에 인터페이스 또는 타입 별칭 사용
	- [x] 타입 별칭 또는 유틸리티 타입을 사용해 타입 복잡성 최소화
	- [x] `declare`를 사용하여 타입을 오버라이드하거나 확장합니다. (예: `req.user`)

## 개발 환경 설정

- [x] `ts-node` 를 사용해 `.ts` 코드를 바로 실행할 수 있는 npm script 생성 (예: `npm run dev`)
- [x] `nodemon`을 사용해 `.ts` 코드가 변경될 때마다 서버가 다시 실행되는 npm script 생성 (예: `npm run dev`)

# 심화 요구 사항

- [x] Layered Architecture 적용
	- [x] Controller, Service, Repository로 나누어 코드 리팩토링
	- [x] 계층 사이에서 데이터를 주고 받을 때 DTO 활용



## 멘토에게
- declare 사용을 조금 다르게 했습니다. 'lib/types.ts'에서 AuthRequest를 import하도록 만들었어요. 이렇게 하는게 간단하다 판단했는데, 틀린 것이라면 말씀해주세요.

unzzang committed:
sprint5 내용 추가

unzzang committed:
파일 및 폴더 정리

kes2020 commented:
declare 사용을 조금 다르게 했습니다. 'lib/types.ts'에서 AuthRequest를 import하도록 만들었어요. 이렇게 하는게 간단하다 판단했는데, 틀린 것이라면 말씀해주세요.

상운님께서 적용하신방식은 기능적으로 문제가 없습니다. 그러나 controller단에서 req에대해서 AuthRequest로 타입캐스팅이필요할것같아서 그냥 Request객체에 필요한 값을 추가 해주시는편이 좀더 좋을 것 같아요~! 
다만 Express에서의 일반적인 타입확장은 수업시간에 배우신대로 types에 적용하셔서 글로벌로 적용해주시는 방법입니다~!

unzzang committed:
refactor: 컨트롤러 및 라우터 구조를 클래스 기반으로 통일

userController 및 authController를 기존 함수 방식에서 클래스 기반으로 리팩토링했습니다.

productController와 동일한 의존성 주입(DI) 패턴을 적용하여 프로젝트 전체의 코드 일관성을 향상시켰습니다.

이에 따라 userRouter, authRouter의 인스턴스 생성 및 라우트 연결 방식을 수정했습니다.

unzzang committed:
declare 관련 내용수정, errorHandler 관련내용 추가

unzzang commented:
@kes2020 
은솔님! declare 관련한 내용 수정했습니다. errorHandler도 조금 더 세분화해서 대응할 수 있도록 추가했습니다.

unzzang committed:
잘못 추가된 파일 삭제

unzzang committed:
chore: Remove .DS_Store files

unzzang committed:
.DS_Store 파일 삭제

unzzang committed:
sprint2, sprint4에 있는 .gitignore 수정

kes2020 commented on src/controllers/articleController.ts:
```diff
@@ -0,0 +1,67 @@
+   * 게시글 목록
+   */
+  public getArticles = async (req: Request, res: Response) => {
+    const { skip, take } = (req as any).pagination;
```
skip ,take을 any로 사용하시는것보다. 
pagination을  구현하는데 필요한 값들을 authRequest구현하셨듯이 pagenationRequest를 interface선언하셔서 사용하시거나

```
export interface PaginatedRequest extends Request {
  pagination: {
    skip: number;
    take: number;
  };
}
```

```
public getArticles = async (req: PaginatedRequest, res: Response) => {
  const { skip, take } = req.pagination; 
  const articles = await this.articleService.findArticles({ skip, take });
  res.status(200).json(articles);
};
```

또 다른 방법으로는 DTO를 선언해서 사용하는방법도 있을 것 같습니다!

kes2020 commented on src/services/productService.ts:
```diff
@@ -0,0 +1,109 @@
+  async findProductById(id: Product['id']) {
+    const product = await this.productRepository.findProductById(id);
+    if (!product) {
+      const error = new Error('상품을 찾을 수 없습니다.');
```
이부분에서 상운님이 작성하신 custom error를 사용하시면 좋을것같습니다. 

ex). NotFoundError()

kes2020 commented:
router.js 파일들에 불필요한 import들이 선언 되어있습니다. 불필요한 import는 제거 해 주시면 좋을 것 같아요~! 
custom error handler들을 작성하셨으니 로직구현부에 해당 customerror를 사용 해 주시면 유지보수하기 좋습니다!

kes2020 commented on src/controllers/authController.ts:
```diff
@@ -0,0 +1,64 @@
+   * 로그아웃(logout)
+   */
+  public logout = async (req: Request, res: Response) => {
+    const userId = (req as any).user.id;
```
declare로 req타입이 보장되므로 

req.user!.id; 로 선언 해 주시면 됩니다.
